Basic interface for bottom-up search

THinnerichs opened the pull request:
Following the design given in: https://github.com/Herb-AI/HerbSearch.jl/pull/107#issuecomment-2658582741

- [x] Test case where some programs are not added to the bank, e.g. observational equivalence.
- [x] Address to include types
- [x] Test with grammars that have multiple types
- [x] Make iterators work with the `@programiterator` macro
- [x] Make work with `UniformTrees`
- [x] Fix cases with arity>1
- [x] Check if all necessary functions are listed in the interface description at the top
- [ ] Check that concrete types (`AcessAddress`) are not used in default method implementations. If they are, it forces the user to rewrite most of the functions if they want to use a different address.
- [ ] Review interface: naming? Are all types necessary? Is the combine-stage state info needed? If so, where?
- [x] Concrete example of rewriting an existing bottom-up iterator from the literature
- [ ] Concrete example of including a constraint (`Ordered` might be a difficult example, `Contains` might be easier) in bottom-up

Commit:
init

Commit:
init refines

Commit:
finished skeleton

Commit:
some fixes

Commit:
full working (?) version

Commit:
add_to_bank fix

Commit:
add_to_bank

ReubenJ committed:
Add basic bottom-up test

ReubenJ committed:
Add failing test for combine

ReubenJ committed:
Fix `combine` in bottom-up

ReubenJ committed:
Add missing comma

ReubenJ committed:
Quick fix for duplicate programs

ReubenJ committed:
Add more bottom up tests

THinnerichs commented:
- [ ] Make iterators work with the `@programiterator` macro:
```julia
julia> @programiterator MyBU(bank) <: BottomUpIterator
MyBU

julia> iter = MyBU(g, :Int, nothing; max_depth=5)
MyBU(GenericSolver(1: Int = 1
2: Int = 2
3: Int = Int + Int
, SolverState(hole[Bool[1, 1, 1]], Set{AbstractLocalConstraint}(), true), DataStructures.PriorityQueue{AbstractLocalConstraint, Int64, Base.Order.ForwardOrdering}(), nothing, false, 9223372036854775807, 5), nothing)

julia> progs = Set(Iterators.take(iter, 6))
ERROR: setfield!: immutable struct of type MyBU cannot be changed
```

sebdumancic commented:
BU doesn't yet work with UniformTrees, but it should be easy to get there now.

Secondly, what was the conclusion: do we want to keep `@programiterator` to move to a more julian way with interfaces?

If we stay with the macro, can it switch to mutable structure? Because that is important for bottom-up search.

ReubenJ committed:
Make `combine` lazier, use `Tuple`s in `CombineAddress`

ReubenJ committed:
Test that BUS == DFS

ReubenJ committed:
Fix search comparison to not hide duplicates

ReubenJ committed:
Make stub functions actual stubs, not empty

ReubenJ committed:
Fix argument names for BUState

sebdumancic commented:
Any suggestions for the interface function names? Especially the ones that manipulate `BottomUpState`, they are very uninspired and potentially uninformative.

ReubenJ committed:
Fix duplicate additions to bank in BUS

ReubenJ committed:
Ensure syntactical duplicates are never added to the bank

ReubenJ committed:
Add get_bank for BUS

ReubenJ commented:
We can remove the mutable keyword for BUS iterators if we switch the interface slightly. What if we remove `create_bank!` (that can be done in the iterator's constructor, removing the need for mutability), and rename `populate_bank!` -> `initialize_bank` (it could `empty!` the bank and add primitives)?

ReubenJ commented:
> Secondly, what was the conclusion: do we want to keep `@programiterator` to move to a more julian way with interfaces?

I think the conclusion was to keep it for now. The interface approach will be nice eventually, but it'll take some work.

> If we stay with the macro, can it switch to mutable structure? Because that is important for bottom-up search.

As far as I can see, it isn't important. Mutating the items _in_ the bank does not require mutability of the structure. The only reason why it's needed now is because of the `iter.bank = ...` assignment in `create_bank!`.

ReubenJ commented:
Either way, I just noticed we have support for `mutable` structs with the `@programiterator` macro.

ReubenJ committed:
Make BUS work with `@programiterator`

ReubenJ commented:
> * [x]  Make iterators work with the `@programiterator` macro:

See the latest commit: this works now. I updated the example BU iterator in the tests to use `@programiterator` now too.

sebdumancic commented:
> We can remove the mutable keyword for BUS iterators if we switch the interface slightly. What if we remove `create_bank!` (that can be done in the iterator's constructor, removing the need for mutability), and rename `populate_bank!` -> `initialize_bank` (it could `empty!` the bank and add primitives)?

This actually makes more sense than the original design. I somehow convinced myself that the original design could be more type stable.

ReubenJ committed:
Use `UniformHole`s

ReubenJ commented:
We can get rid of the `current_uniform_iterator::Union{UniformIterator,Nothing}` field in the `GenericBUState` (it can live in the solver), but I'm pushing now as tests pass

ReubenJ commented:
We also assume grammars with a single type, there is a TODO in there mentioning the change needed, but addressing will need to include this too

ReubenJ committed:
Implementation broken with arity>1

Added tests to show this

ReubenJ committed:
Restructure bottom-up tests

ReubenJ committed:
Fix bottom-up terminating early

ReubenJ committed:
Fix bottom-up derivation heuristic

ReubenJ committed:
Add test grammar for bottom-up with multiple types

ReubenJ commented:
Updated. Also, added a (failing) test with multiple types.

ReubenJ commented:
Need https://github.com/Herb-AI/HerbConstraints.jl/issues/86 to make search over a grammar with >1 types work.

ReubenJ committed:
Support multiple types with BUS

Currently, exploring duplicate programs when the grammar
includes an operator like `Int * Int`

ReubenJ committed:
Add more bottom up tests

ReubenJ committed:
Fix bottom up enumeration duplicates

ReubenJ committed:
Add/test example of bottom-up with obs. equiv.

ReubenJ committed:
Avoid uneccessary iterations over uniform trees

ReubenJ committed:
Fix optimization to avoid extra uniform iter. iterations

ReubenJ committed:
Make iteration iterative, and remove any recursion

This commit breaks comparison tests with DFS

ReubenJ committed:
Add missing depth check

ReubenJ committed:
Re-enable all tests

ReubenJ committed:
Fix div. and conquer tests

ReubenJ committed:
Move all tests inside of one testset

ReubenJ committed:
Increase test depth to 4

ReubenJ committed:
Switch to depth, fix address indices

ReubenJ committed:
Remove obs. equiv. for now

ReubenJ committed:
Remove hard-coded depth parameter

ReubenJ committed:
Remove extra `@show` to clean up output

ReubenJ committed:
Clean up BUS: interface, implementation, tests

ReubenJ committed:
Simplify iterate

ReubenJ committed:
Add new `retrieve(bank, addr)` method

Update docstrings for all methods of `retrieve`

ReubenJ committed:
Remove unused `create_bank` function

ReubenJ committed:
Update bottom up docstrings

ReubenJ committed:
Resolve `retrieve` ambiguities

ReubenJ committed:
Fix doctests for `<: AbstractAddress`es

ReubenJ committed:
Update docstrings, use generated signatures

ReubenJ committed:
Use `return_type` instead of `get_type`

ReubenJ committed:
Re-enable all tests

ReubenJ committed:
Fix exports, formatting

ReubenJ committed:
Introduce methods in place of field accesses

Further clears up Addresses by breaking up their `addr` fields

ReubenJ committed:
Add doctest testing, fix some docstrings

ReubenJ committed:
Fix doctests

ReubenJ committed:
Update tests for the explicit structure bank

ReubenJ committed:
Revert AccessAddress rename, switch to explicit bank structure

ReubenJ committed:
Add, fix up docstrings

ReubenJ committed:
Some doctest/docstring fixes

ReubenJ commented on src/bottom_up_iterator.jl:
```diff
@@ -0,0 +1,580 @@
+using HerbCore: RuleNode
+import HerbGrammar.return_type
+
+"""
+    abstract type BottomUpIterator <: ProgramIterator
+
+A type of iterator that maintains a bank of (executable) programs and iteratively explores
+larger programs by combining the ones in the bank.
+
+Like other [`ProgramIterator`](@ref)s, `BottomUpIterator`s enumerate programs from a
+context-free grammar starting at `Symbol` `sym` with respect to the grammar up to a given
+depth and a given size.
+
+The central concept behind this interface is an _address_ ([`AbstractAddress`](@doc)).
+Addresses are simply pointers into the bank. Since the bank is often a nested container (ex:
+`Dict{Key,Dict{Key,...}}`), the addresses usually have multiple components. The first
+components indexes the top-most level of the bank, the second indexes, the next level, etc.
+Program combinations can be expressed as a special address ([`CombineAddress`](@ref)) that
+contains one or more programs and an operator. For example: combining the programs `1 + x`
+and `x * x` with the `+` operator results in the program `(1 + x) + (x * x)`.
+
+The interface for bottom-up iteration is defined as follows.
+
+- [`get_bank`](@ref): get the iterator's bank
+- [`populate_bank!`](@ref): initialize the bank with the terminals from the grammar
+  and return the resulting [`AccessAddress`](@ref)es
+- [`combine`](@ref): combine the existing programs in the bank into new, more complex
+  programs via [`CombineAddress`](@ref)es
+- [`add_to_bank!`](@ref): possibly add a program created by the [`combine`](@ref) step to
+  the bank
+- [`retrieve`](@ref): retrieve the program from the bank given an [`AbstractAddress`](@ref)
+
+A generic implementation ([`SizeBasedBottomUpIterator`](@ref)) is given with a bank that is
+indexed based on the program size, meaning that each level of the bank has programs
+represented by the same number of nodes. Because the implementation works using an arbitrary
+grammar, the bank also must be indexed on the type of the programs to allow the
+[`combine`](@ref) step to avoid constructing programs that do not adhere to the grammar.
+"""
```
@sebdumancic for you review, please take a look at this overview

ReubenJ reviewed (approved):
From my side, this is good to go. If you can take a pass over this when you have a chance @sebdumancic, that'd be great. After that, I'd say let's merge.

THinnerichs committed:
Rename getters to be more recognizable; add more documentation

THinnerichs committed:
Added default functions for get_max_size/max_depth for program iterator; updated limit for SizeBasedBUS

THinnerichs committed:
Update Cost-based search

THinnerichs committed:
Update for checking depth and size for all iterators

THinnerichs committed:
Shape-based cost-based search

THinnerichs committed:
Update cost-based search for two-part bank

THinnerichs committed:
Add implementation for Cost-based BUS; Add generalized test suite

THinnerichs committed:
Refactor size-based and depth-based BUS, add horizon formulation; add tests and performant optimizations

THinnerichs committed:
Refactor shape and cost-based search into one structure

THinnerichs committed:
Iteration time to 7.28seconds

THinnerichs committed:
Iteration time to 6.8

THinnerichs committed:
5.5 seconds

THinnerichs committed:
Clean up build_cost_cross_product

THinnerichs committed:
Add mono-program cost-based BUS

THinnerichs committed:
Add constraint checking + optimizations

THinnerichs committed:
Fix observational equivalence for CB-BUS

THinnerichs committed:
Add documentation; clean-up code; standardize interface

THinnerichs committed:
Add tests, all passing

THinnerichs committed:
Add parametric types; add shape-based combine for CB-BUS; update observational equivalence computation

THinnerichs committed:
Fix Parametric type issue

THinnerichs committed:
Merge branch 'dev-bu-basic-interface' into feat/cost-based_BUS

THinnerichs committed:
Merge pull request #163 from Herb-AI/feat/cost-based_BUS

Add cost-based search with OE, fix generic BUS

THinnerichs committed:
Fix tests using observational equivalence

THinnerichs committed:
Merge branch 'master' into dev-bu-basic-interface

THinnerichs committed:
Fix deprecated priority queue functions (enqueue, dequeue, peek); fix exports; fix merge error in program_iterator

ReubenJ committed:
Merge branch 'master' into dev-bu-basic-interface

ReubenJ committed:
fix doctests
